File: packages/rust/client/src/txbuilder/withdraw.rs
use crate::instructions::{Withdraw, WithdrawInstructionArgs};
use crate::utils::get_create_ata_instructions;
use crate::{get_lending_position_address, get_tuna_config_address, get_vault_address};
use solana_program::instruction::Instruction;
use solana_program::pubkey::Pubkey;
use spl_associated_token_account::get_associated_token_address;

pub fn withdraw_instructions(authority: &Pubkey, mint: &Pubkey, funds: u64, shares: u64) -> Vec<Instruction> {
    let authority_ata_instructions = get_create_ata_instructions(&mint, authority, authority, &spl_token::ID, 0);

    let mut instructions = vec![];
    instructions.extend(authority_ata_instructions.create);
    instructions.push(withdraw_instruction(authority, mint, funds, shares));
    instructions.extend(authority_ata_instructions.cleanup);

    instructions
}

pub fn withdraw_instruction(authority: &Pubkey, mint: &Pubkey, funds: u64, shares: u64) -> Instruction {
    let tuna_config_address = get_tuna_config_address().0;
    let vault_address = get_vault_address(&mint).0;
    let lending_position_address = get_lending_position_address(&authority, &mint).0;

    let authority_ata = get_associated_token_address(&authority, &mint);
    let vault_ata = get_associated_token_address(&vault_address, &mint);

    let ix_builder = Withdraw {
        authority: *authority,
        authority_ata,
        mint: *mint,
        tuna_config: tuna_config_address,
        vault: vault_address,
        vault_ata,
        lending_position: lending_position_address,
        token_program: spl_token::ID,
        associated_token_program: spl_associated_token_account::ID,
    };

    ix_builder.instruction(WithdrawInstructionArgs { funds, shares })
}

File: packages/ts/client/src/txbuilder/withdraw.ts
import {
  Account,
  Address,
  GetAccountInfoApi,
  GetMultipleAccountsApi,
  IInstruction,
  Rpc,
  TransactionSigner,
} from "@solana/kit";
import { MEMO_PROGRAM_ADDRESS } from "@solana-program/memo";
import { fetchMaybeMint, findAssociatedTokenPda, Mint } from "@solana-program/token-2022";

import {
  getCreateAtaInstructions,
  getLendingPositionAddress,
  getLendingVaultAddress,
  getTunaConfigAddress,
  getWithdrawInstruction,
} from "../index.ts";

export async function withdrawInstructions(
  rpc: Rpc<GetAccountInfoApi & GetMultipleAccountsApi>,
  authority: TransactionSigner,
  mintAddress: Address,
  funds: bigint,
  shares: bigint,
): Promise<IInstruction[]> {
  const instructions: IInstruction[] = [];

  const mint = await fetchMaybeMint(rpc, mintAddress);
  if (!mint.exists) throw new Error("Mint account not found");

  // Add create user's token account instruction if needed.
  const createUserAtaInstructions = await getCreateAtaInstructions(
    authority,
    mint.address,
    authority.address,
    mint.programAddress,
  );
  instructions.push(...createUserAtaInstructions.init);

  // Add withdraw instruction
  const ix = await withdrawInstruction(authority, mint, funds, shares);
  instructions.push(ix);

  // Close WSOL accounts if needed.
  instructions.push(...createUserAtaInstructions.cleanup);

  return instructions;
}

export async function withdrawInstruction(
  authority: TransactionSigner,
  mint: Account<Mint>,
  funds: bigint,
  shares: bigint,
): Promise<IInstruction> {
  const tunaConfig = (await getTunaConfigAddress())[0];
  const lendingPosition = (await getLendingPositionAddress(authority.address, mint.address))[0];

  const vault = (await getLendingVaultAddress(mint.address))[0];
  const vaultAta = (
    await findAssociatedTokenPda({
      owner: vault,
      mint: mint.address,
      tokenProgram: mint.programAddress,
    })
  )[0];

  const authorityAta = (
    await findAssociatedTokenPda({
      owner: authority.address,
      mint: mint.address,
      tokenProgram: mint.programAddress,
    })
  )[0];

  return getWithdrawInstruction({
    authority,
    authorityAta,
    lendingPosition,
    mint: mint.address,
    tunaConfig,
    vault,
    vaultAta,
    tokenProgram: mint.programAddress,
    memoProgram: MEMO_PROGRAM_ADDRESS,
    funds,
    shares,
  });
}
